proposals/2024-09-25_metadata-labels.md

```diff
@@ -0,0 +1,186 @@
+
+#### Milestone 1: Feature flag for adding labels
+
+Add a feature flag: `--enable-feature=identifying-type-and-unit`. When enabled, `__type__` and `__unit__` labels are added when receiving OTLP or PRW 2.0, or when scraping the text, OM, or proto formats. Implement any changes required to allow relabeling `__type__` and `__unit__` in `metric_relabel_configs`.
+
+#### Milestone 2: UI and PromQL changes
+
+During this stage, implement the UI and PromQL changes above. Iterate based on feedback. Changes should have no effect unless the identifying-type-and-unit feature flag is enabled.
+
+#### Milestone 3: Add NoNameChanges option for OTLP translation
+
+Add an option, `NoNameChanges` for the OTLP translation strategy. When enabled, it disables UTF-8 sanitization and the addition of suffixes. In the documentation for this option, recommend that `--enable-feature=identifying-type-and-unit` is enabled.
+
+## Alternatives
+
+### “Real” Type and Unit suffixes in **name**
+
+Similar to suffixes of _<unit> and _<type/total> but make it an explicit suffix using a delimiter not currently permitted in metric names. Specifying suffixes is optional when querying for a metric. When the type or unit suffix is omitted from a query, it would (design TBD) return results which include any type or unit suffix which exists for that name.
+
+NOTES:
+
+* `~` for units and `.` for type is just one example, there might be better operators/characters to use.
+* This proposal is fully compatible with the proposal above. `http_request_duration{__unit__="seconds", __type__="histogram"}` could just be another syntax for `http_request_duration~seconds.histogram`. It would make it much easier to add units and/or types to the metric name, so it would address the concern that you cannot see the unit and type anymore by looking at a PromQL expression without supporting tooling. If we allowed `.total` as an alias of `.counter`, we would have very little visible change. `http_requests_total` would become `http_requests.total`.
+
+Writing queries that include the type and unit would be recommended as a best-practice by the community.
+
+For example:
+
+* Querying for `foo.histogram` would return results that include both `foo~seconds.histogram` and `foo~milliseconds.histogram`.
+* Querying for `foo~seconds` would return results that include both `foo~seconds.histogram` and `foo~seconds.counter`.
+* Querying for `http_server_duration` would return results that include both `foo~seconds.histogram` and `foo~milliseconds.counter`.
+* Querying for an OpenTelemetry metric, such as `http.server.duration`, with suffixes would require querying for `”http.server.duration”~seconds.histogram`. Note that suffixes are outside of quotes.
+
+This solution is not chosen because:
+
+* Adding suffixes outside of quotes looks strange: `{“http.server.duration”~seconds.histogram}`.
+  * Alternatives: `{“http.server.duration”}~seconds.histogram` or `{“http.server.duration”}{seconds,histogram}`
```

Comment:
I would see `{"http.server.duration"}~seconds.histogram` as the most "obvious" syntax.

Reply:
I think this deserves to go through the hands of UX researchers :) 

I would be very confused to see tildes and dots suddenly in my metric names. I see this as a source of confusion, as there are more things that I need to learn to use Prometheus effectively.

Reply:
The usage is optional, so at least you don't _have_ to learn in to write these.

The problem is that the core of this proposal, namely putting type and unit into labels rather than give hints via naming conventions, also needed to go through the hands of UX researchers. The proposed option is in fact an attempt to address UX concerns.

Reply:
Should we set an explicit non-goal of designing a short syntax for specifying type and unit? that feels like a "nice-to-have" as opposed to a requirement, and it seems non-blocking on the overall work.

Reply:
I already did that I believe.

Reply:
Well.. now - I pushed to a wrong branch before https://github.com/prometheus/proposals/pull/39/commits/7b703a6f35a3db26d5a2cd9e2878dbd8d7ada6b8